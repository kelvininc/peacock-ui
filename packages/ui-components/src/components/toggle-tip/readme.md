# _kv-toggle-tip_



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-toggle-tip text="Toggle tip">
	<kv-icon slot="open-element-slot" name={EIconName.Info}/>
	<kv-action-button [type]="EActionButtonType.Primary" slot="content-slot">Learn More</kv-action-button>
</kv-toggle-tip>

<!-- With Fixed Position -->
<kv-toggle-tip text="Toggle tip" [position]="ETooltipPosition.Left">
	<kv-icon slot="open-element-slot" name={EIconName.Info}/>
	<kv-action-button [type]="EActionButtonType.Primary" slot="content-slot">Learn More</kv-action-button>
</kv-toggle-tip>

<!-- With Allowed Position -->
<kv-toggle-tip text="Toggle tip" [allowedPositions]="[ETooltipPosition.Top, ETooltipPosition.Bottom]">
	<kv-icon slot="open-element-slot" name={EIconName.Info}/>
	<kv-action-button [type]="EActionButtonType.Primary" slot="content-slot">Learn More</kv-action-button>
</kv-toggle-tip>
```


### React

```tsx
import React from 'react';

import { KvToggleTip, KvIcon, KvActionButton, EIconName, ETooltipPosition, EActionButtonType } from '@kelvininc/react-ui-components';

export const ToggleTipExample: React.FC = () => (
  <>
    {/*-- Default --*/}
	<KvToggleTip text="Toggle tip">
		<KvIcon slot="open-element-slot" name={EIconName.Info}/>
		<KvActionButton slot="content-slot" type={EActionButtonType.Primary}>Learn More</KvActionButton>
	</KvToggleTip>

	{/*-- With Fixed Position --*/}
	<KvToggleTip text="Toggle tip" position={ETooltipPosition.Left}>
		<KvIcon slot="open-element-slot" name={EIconName.Info}/>
		<KvActionButton slot="content-slot" type={EActionButtonType.Primary}>Learn More</KvActionButton>
	</KvToggleTip>

	{/*-- With Allowed Positions --*/}
	<KvToggleTip text="Toggle tip" allowedPositions={[ETooltipPosition.Top, ETooltipPosition.Bottom]}>
		<KvIcon slot="open-element-slot" name={EIconName.Info}/>
		<KvActionButton slot="content-slot" type={EActionButtonType.Primary}>Learn More</KvActionButton>
	</KvToggleTip>
  </>
);
```



## Properties

| Property           | Attribute    | Description                                                                                | Type                                                                                                                                                                                                                                                                                                                                             | Default     |
| ------------------ | ------------ | ------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------- |
| `allowedPositions` | --           | (optional) Array of allowed positions of toggle tip (if defined the 'position' is ignored) | `ETooltipPosition[]`                                                                                                                                                                                                                                                                                                                             | `undefined` |
| `arrowElement`     | --           | (optional) html arrow element ref                                                          | `HTMLElement`                                                                                                                                                                                                                                                                                                                                    | `null`      |
| `containerElement` | --           | (optional) html toggle tip container ref                                                   | `HTMLElement`                                                                                                                                                                                                                                                                                                                                    | `null`      |
| `disabled`         | `disabled`   | (optional) if true it will disable clicks to open toggle tip                               | `boolean`                                                                                                                                                                                                                                                                                                                                        | `false`     |
| `isFixed`          | `is-fixed`   | (optional) if true it will ignore outside clicks to close the toggle tip                   | `boolean`                                                                                                                                                                                                                                                                                                                                        | `false`     |
| `isOpen`           | `is-open`    | (optional) Set open state of toggle tip, default false                                     | `boolean`                                                                                                                                                                                                                                                                                                                                        | `false`     |
| `openElement`      | --           | (optional) html toggle tip opening element ref                                             | `HTMLElement`                                                                                                                                                                                                                                                                                                                                    | `null`      |
| `position`         | `position`   | (optional) Position of toggletip                                                           | `ETooltipPosition.Bottom \| ETooltipPosition.BottomEnd \| ETooltipPosition.BottomStart \| ETooltipPosition.Left \| ETooltipPosition.LeftEnd \| ETooltipPosition.LeftStart \| ETooltipPosition.Right \| ETooltipPosition.RightEnd \| ETooltipPosition.RightStart \| ETooltipPosition.Top \| ETooltipPosition.TopEnd \| ETooltipPosition.TopStart` | `undefined` |
| `text`             | `text`       | (optional) Text of toggletip                                                               | `string`                                                                                                                                                                                                                                                                                                                                         | `undefined` |
| `withArrow`        | `with-arrow` | (optional) if true it will render an arrow pointing to the opening element                 | `boolean`                                                                                                                                                                                                                                                                                                                                        | `false`     |


## Events

| Event             | Description                                   | Type                   |
| ----------------- | --------------------------------------------- | ---------------------- |
| `openStateChange` | Emitted when the dropdown opens state changes | `CustomEvent<boolean>` |


## Shadow Parts

| Part                        | Description                  |
| --------------------------- | ---------------------------- |
| `"toggle-tip-container"`    | The toggle tip container.    |
| `"toggle-tip-slot-content"` | The toggle tip slot content. |


## CSS Custom Properties

| Name                               | Description           |
| ---------------------------------- | --------------------- |
| `--toggle-tip-container-max-width` | toggle tip max width. |
| `--toggle-tip-container-width`     | toggle tip width.     |
| `--toggletip-z-index`              | toggle tip z-index.   |


----------------------------------------------


