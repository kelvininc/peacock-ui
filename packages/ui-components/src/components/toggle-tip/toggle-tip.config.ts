import { ComputePositionConfig } from '@floating-ui/dom';

import { AutoPlacementOptions } from '@floating-ui/dom';
import { DetectOverflowOptions as CoreDetectOverflowOptions } from '@floating-ui/dom';

export const DEFAULT_TOOLTIP_CONFIG: Partial<ComputePositionConfig> = {
	strategy: 'fixed'
};

export const DEFAULT_AUTO_PLACEMENT_CONFIG: Partial<AutoPlacementOptions & CoreDetectOverflowOptions> = {
	padding: 5
};
