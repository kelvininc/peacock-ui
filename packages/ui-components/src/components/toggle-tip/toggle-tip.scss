@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --toggletip-z-index: toggle tip z-index.
	 * @prop --toggle-tip-container-width: toggle tip width.
	 * @prop --toggle-tip-container-max-width: toggle tip max width.
	 */

	--toggletip-z-index: 1000;
	--toggle-tip-container-width: fit-content;
	--toggle-tip-container-max-width: 292px;
}

.toggle-tip-open-element-wrapper {
	width: fit-content;
	height: fit-content;
}

.toggle-tip-container {
	height: fit-content;
	width: var(--toggle-tip-container-width);
	max-width: var(--toggle-tip-container-max-width);
	word-wrap: break-word;
	color: kv-color('neutral-0');
	background-color: kv-color('neutral-6');
	box-shadow: 0 8px 16px -4px kv-color('neutral-6', 'base', 0.25), 0 0 1px 0 kv-color('neutral-3', 'base', 0.31);
	user-select: none;
	border-radius: 2px;
	padding: $spacing-3x;
	position: fixed;
	display: none;
	z-index: var(--toggletip-z-index);
}

.toggle-tip-text {
	@include kv-font-span-semibold();
	color: kv-color('neutral-2');
}

.toggle-tip-arrow {
	position: absolute;
	background: kv-color('neutral-6');
	width: 10px;
	height: 10px;
	transform: rotate(45deg);
}
