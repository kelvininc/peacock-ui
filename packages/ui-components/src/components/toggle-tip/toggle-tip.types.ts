import { EventEmitter } from '@stencil/core';
import { ETooltipPosition } from '../../utils/types/components';

export interface IToggleTip {
	/** (optional) Text of toggletip */
	text?: string;
	/** (optional) Position of toggletip */
	position?: ETooltipPosition;
	/** (optional) Array of allowed positions of toggle tip (if defined the 'position' is ignored) */
	allowedPositions?: ETooltipPosition[];
	/** (optional) Set open state of toggle tip, default false */
	isOpen?: boolean;
	/** (optional) if true it will ignore outside clicks to close the toggle tip */
	isFixed?: boolean;
	/** (optional) if true it will render an arrow pointing to the opening element */
	withArrow?: boolean;
	/** (optional) if true it will disable clicks to open toggle tip */
	disabled?: boolean;
	/** (optional) html toggle tip opening element ref */
	openElement?: HTMLElement;
	/** (optional) html toggle tip container ref */
	containerElement?: HTMLElement;
	/** (optional) html arrow element ref */
	arrowElement?: HTMLElement;
}

export interface IToggleTipEvents {
	/** Emitted when the dropdown opens state changes */
	openStateChange: EventEmitter<boolean>;
}
