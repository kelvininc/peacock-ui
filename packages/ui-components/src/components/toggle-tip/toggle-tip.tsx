import { autoPlacement, ComputePositionConfig, Middleware } from '@floating-ui/dom';
import { Component, Host, h, Prop, Element, Listen, Event, EventEmitter } from '@stencil/core';
import { isEmpty } from 'lodash-es';
import { CustomCssClass, ETooltipPosition } from '../../utils/types/components';
import { DEFAULT_AUTO_PLACEMENT_CONFIG, DEFAULT_TOOLTIP_CONFIG } from './toggle-tip.config';
import { IToggleTip, IToggleTipEvents } from './toggle-tip.types';
import { getClassMap } from '../../utils/css-class.helper';
import { didClickOnElement } from '../../utils/mouse-event.helper';
import { TOGGLE_TIP_Z_INDEX } from '../../globals/config';
import { mergeComputePositionConfigs } from '../../utils/floating-ui.helper';

@Component({
	tag: 'kv-toggle-tip',
	styleUrl: 'toggle-tip.scss',
	shadow: false
})
export class KvToggleTip implements IToggleTip, IToggleTipEvents {
	/** @inheritdoc */
	@Prop({ reflect: true }) text?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) position?: ETooltipPosition;
	/** @inheritdoc */
	@Prop({ reflect: true }) allowedPositions?: ETooltipPosition[];
	/** @inheritdoc */
	@Prop({ reflect: false }) options?: Partial<ComputePositionConfig> = DEFAULT_TOOLTIP_CONFIG;
	/** @inheritdoc */
	@Prop({ reflect: true, mutable: true }) isOpen?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) isFixed?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) withArrow?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) customClass?: CustomCssClass = '';

	/** @inheritdoc */
	@Event() openStateChange: EventEmitter<boolean>;

	/** The Host's element reference */
	@Element() el: HTMLKvToggleTipElement;

	@Listen('click', { target: 'window' })
	checkForClickOutside(event) {
		if (!this.isFixed) {
			// Check if clicked inside the toggle tip
			if (didClickOnElement(this.portal, event) || didClickOnElement(this.openElement, event)) {
				return;
			}
			if (this.isOpen) {
				this.isOpen = false;
				this.openStateChange.emit(false);
			}
		}
	}

	private portal: HTMLElement;
	private openElement: HTMLElement;

	private getOptions = (): Partial<ComputePositionConfig> => {
		const placement = isEmpty(this.allowedPositions) ? this.position : undefined;
		const middleware: Array<Middleware> = [];

		if (!isEmpty(this.allowedPositions) && isEmpty(placement)) {
			middleware.push(
				autoPlacement({
					...DEFAULT_AUTO_PLACEMENT_CONFIG,
					allowedPlacements: this.allowedPositions
				})
			);
		}

		return mergeComputePositionConfigs({ placement, middleware }, this.options);
	};

	private onButtonClick = () => {
		if (this.disabled) {
			return;
		}

		this.isOpen = !this.isOpen;
		this.openStateChange.emit(this.isOpen);
	};

	disconnectedCallback() {
		// Requires deleting portal from outside KvPortal because KvPortal is moved to global context
		// and would only be destroyed when the global context is destroyed.
		this.portal.remove();
	}

	render() {
		return (
			<Host>
				<div ref={el => (this.openElement = el)} onClick={this.onButtonClick} class="toggle-tip-open-element-container">
					<slot name="open-element-slot"></slot>
				</div>
				<kv-portal
					zIndex={TOGGLE_TIP_Z_INDEX}
					ref={el => (this.portal = el)}
					withArrow={this.withArrow}
					animated
					show={this.isOpen}
					reference={this.openElement}
					options={this.getOptions()}
				>
					<kv-tooltip-text class={{ 'toggle-tip-container': true, ...getClassMap(this.customClass) }} text={this.text}>
						{/* Shadow Root should be false to slot work here */}
						<div>
							<slot name="content-slot" />
						</div>
					</kv-tooltip-text>
				</kv-portal>
			</Host>
		);
	}
}
