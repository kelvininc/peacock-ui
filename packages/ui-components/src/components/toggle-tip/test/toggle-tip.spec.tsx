import { SpecPage } from '@stencil/core/internal';
import { KvToggleTip } from '../toggle-tip';
import { newSpecPage } from '@stencil/core/testing';

describe('Toggle Tip (unit tests)', () => {
	let page: SpecPage;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleTip],
				html: `<kv-toggle-tip text="Toggle tip"><kv-icon slot="open-element-slot" name="kv-logo-kelvin"/></kv-toggle-tip>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has a position', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvToggleTip],
				html: `<kv-toggle-tip text="Toggle tip" position="left"><kv-icon slot="open-element-slot" name="kv-logo-kelvin"/></kv-toggle-tip>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
