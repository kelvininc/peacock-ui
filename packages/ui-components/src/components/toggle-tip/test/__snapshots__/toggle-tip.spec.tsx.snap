// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Tip (unit tests) when has a position should match the snapshot 1`] = `
<kv-toggle-tip custom-class="" position="left" text="Toggle tip">
  <!---->
  <div class="toggle-tip-open-element-container">
    <kv-icon name="kv-logo-kelvin" slot="open-element-slot"></kv-icon>
  </div>
  <kv-portal animated="" zindex="9006">
    <kv-tooltip-text class="toggle-tip-container" text="Toggle tip">
      <div></div>
    </kv-tooltip-text>
  </kv-portal>
</kv-toggle-tip>
`;

exports[`Toggle Tip (unit tests) when uses default props should match the snapshot 1`] = `
<kv-toggle-tip custom-class="" text="Toggle tip">
  <!---->
  <div class="toggle-tip-open-element-container">
    <kv-icon name="kv-logo-kelvin" slot="open-element-slot"></kv-icon>
  </div>
  <kv-portal animated="" zindex="9006">
    <kv-tooltip-text class="toggle-tip-container" text="Toggle tip">
      <div></div>
    </kv-tooltip-text>
  </kv-portal>
</kv-toggle-tip>
`;
